### ISSUE TITLE
Possible invalid memory access?

### ISSUE THREAD

**Issue by cpennington** [HuwCampbell/grenade#37]
I think the `im2col` (and maybe `col2im`) can potentially read uninitialized memory: Running `Grenade.Layers.Internal.Convolution.im2col 10 10 2 2 $ (5><6) [0..]` gives a bunch of small float values, which seem likely to be because of reading invalid memory locations.

**Comment by HuwCampbell** [comment]
`im2col` isn't a total function, it only works where the kernel and stride "fit" for the input the matrix. So yes, it does look like that's what's happening.

This "fitting" constraint is captured in the proofs going into the convolution and deconvolution layers. I found that if I tried to push these through to the `im2col` function it lead to a considerable slowdown, so put them in an internal function set instead.

It could probably be better documented, and maybe even just stop with error inside `im2col` if these constraints are violated.
